Fixes issue #7109

jgold6 opened the pull request:
### Description of Change ###

Changes check for iOS version >= 11 to a check to see if the `setNeedsUpdateOfHomeIndicatorAutoHidden` selector is available before calling the `SetNeedsUpdateOfHomeIndicatorAutoHidden` method.

 No automated test created as there should be no difference in behavior except to avoid a crash due to calling an Obj-C selector that may not be available on all devices running iOS >= 11. In this case, it seems that checking that the selector is available is better than checking for the iOS version as it will be a valid check regardless of iOS version, where as just the version check alone is not guaranteeing the availability of this particular selector as devices pre iPhone X do not have a Home Indicator. 

### Issues Resolved ### 

- fixes #7109 

### API Changes ###

None

### Platforms Affected ### 

- iOS


### Behavioral/Visual Changes ###

Users app won't crash when running on an iPhone 6 with iOS version 11 (and potentially other iPhones pre the iPhone X when running iOS 11)

### Before/After Screenshots ### 

Not applicable

### Testing Procedure ###

I was unable to test as I do not have an iPhone 6 running iOS 11. However the native stack trace provided by the customer clearly indicates that the crash is due to calling an Obj-C selector that does not exist.

### PR Checklist ###

- [ ] Has automated tests <!-- (if tests are omitted or manual, state reason in description) -->
- [ ] Rebased on top of the target branch at time of PR
- [x] Changes adhere to coding standard

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/xamarin/Xamarin.Forms?pullRequest=7114) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/xamarin/Xamarin.Forms?pullRequest=7114) before we can accept your contribution.<br/><br/>:x: jgold6 [sign now](https://cla.dotnetfoundation.org/xamarin/Xamarin.Forms?pullRequest=7114)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/xamarin/Xamarin.Forms?pullRequest=7114) it.</sub>

jgold6 commented:
Fixed build failure due to not fully qualifying the Selector type

jfversluis commented:
Could you maybe retarget this to 4.2? Thanks!

jgold6 commented:
So I tried to retarget this (again I am new to this process) and now I see a bunch of other commits in this PR. Not sure if that is expected or not. Let me kn ow if I need to start a new pull request against 4.2 instead of modifying this one as I did.

samhouts commented:
@jgold6 I'd actually like to see this fix ported back all the way to 4.0, but we'll have to do it with multiple PRs because https://github.com/xamarin/Xamarin.Forms/pull/6636 moved ModalWrapper into its own file in 4.1.0. 

Also, I'd like to see the check cached so we're not getting a performance hit every time `ViewDidLoad` runs.

Thanks!!

jgold6 commented:
@samhouts 

The check was not cached before when it was an iOS version check, also I am unsure what mechanism you would want to use to cache the check. 

And, again, forgive me as I am new to creating pull requests for Forms, and not very experienced in creating pull requests period. I see that you tried to rebase this PR to 4.0 and then rebased again to master.

I am guessing that you would want a PR from a branch that is based on 4.0, 4.1, 4.2, and master branch with these changes? Is this correct? Initially I though I could just create the PR against master and then you guys could cherry pick it back to whatever versions you needed it in, but I guess I thought incorrectly.

jfversluis commented:
Hi @jgold6, don't worry! We're here to help.

It's totally up to you, if you want to do it, please do.

With the caching Samantha means to not call `RespondsToSelector(new ObjCRuntime.Selector("setNeedsUpdateOfHomeIndicatorAutoHidden")` each time, but to get it once when the class is created and just use that value from there. That way we just have to access the variable value instead of going through APIs and retrieve the value each time. Since this is something that is not likely to change during runtime that should be fine.

With the PRs I think you can open one for 4.0, 4.1 and retarget this to 4.2. To do the latter, rebase onto upstream/4.2.0, force push and change the branch you want to merge into to 4.2.0.

Let us know if you're up for it, else we'll do it for you :)

Thanks!

hartez commented on Xamarin.Forms.Platform.iOS/PlatformRenderer.cs:
```diff
@@ -91,7 +91,7 @@ public override void ViewDidLoad()
 			base.ViewDidLoad();
 			SetNeedsStatusBarAppearanceUpdate();
-			if (Forms.IsiOS11OrNewer)
+			if (RespondsToSelector(new ObjCRuntime.Selector("setNeedsUpdateOfHomeIndicatorAutoHidden")))
```
The casing for the selector is different in this instance than the others.

I'm wondering if it would make sense to move this into an extension method for UIViewController which does the check and calls `SetNeedsUpdateOfHomeIndicatorAutoHidden()` as appropriate. Rather than duplicating this code in several places.

jgold6 commented:
I am having problems rebasing this branch to upstream/4.2.0. Getting an error about conflicts on a commit: 
```
>git rebase upstream/4.2.0
First, rewinding head to replay your work on top of it...
Applying: bump gitinfo
Applying: RelativeSource Binding (#4375)
Applying: cleanup: corrected spelling mistake (#6794)
Applying: Implemented Padding on WPF Button (#6791)
Applying: cleanup: moved closing brace to new line (#6793)
Applying: [Core] Add ability to show modal prompts (#6714)
Applying: [X] allow x:Type markup as x:DataType value (#6843)
Applying: Label padding (#6299)
Applying: [Tizen] Fix slider thumb color issue (#6785)
Applying: [Enhancement] Implements CharacterSpacing (#5167)
.git/rebase-apply/patch:477: trailing whitespace.

.git/rebase-apply/patch:747: trailing whitespace.

.git/rebase-apply/patch:2010: trailing whitespace.
                {
.git/rebase-apply/patch:2014: trailing whitespace.

.git/rebase-apply/patch:2023: trailing whitespace.
                        return new NSAttributedString(span.Text, font == Font.Default ? null : font.ToUIFont(), fgcolor.ToUIColor(),
warning: 5 lines add whitespace errors.
Using index info to reconstruct a base tree...
M       Xamarin.Forms.Controls/Xamarin.Forms.Controls.csproj
M       Xamarin.Forms.Platform.Android/Renderers/SearchBarRenderer.cs
M       Xamarin.Forms.Platform.UAP/ButtonRenderer.cs
Falling back to patching base and 3-way merge...
Auto-merging Xamarin.Forms.Platform.UAP/ButtonRenderer.cs
Auto-merging Xamarin.Forms.Platform.Android/Renderers/SearchBarRenderer.cs
Auto-merging Xamarin.Forms.Controls/Xamarin.Forms.Controls.csproj
CONFLICT (content): Merge conflict in Xamarin.Forms.Controls/Xamarin.Forms.Controls.csproj
error: Failed to merge in the changes.
Patch failed at 0010 [Enhancement] Implements CharacterSpacing (#5167)
Use 'git am --show-current-patch' to see the failed patch

Resolve all conflicts manually, mark them as resolved with
"git add/rm <conflicted_files>", then run "git rebase --continue".
You can instead skip this commit: run "git rebase --skip".
To abort and get back to the state before "git rebase", run "git rebase --abort".
```

I might close this and just start three new pull requests, one each for 4.0, 4.1 and 4.2

I will fix the casing issue and make a class member to hold the value of the check. 

I, however, don't really see the value of making a method, extension or otherwise, for a two line call? Also this was already duplicated code in all 5 files so I was just following what was done before. All I intended to do was change the existing check from a version check to a selector check.

jgold6 commented:
Closing this. Going to start from scratch.

jgold6 commented:
New pull requests:

XF 4.0: https://github.com/xamarin/Xamarin.Forms/pull/7151
XF 4.1: https://github.com/xamarin/Xamarin.Forms/pull/7149
XF 4.2: https://github.com/xamarin/Xamarin.Forms/pull/7150
